File: oltvi_core/lib/src/services/mock_data_service.dart
import 'dart:math';
import 'package:latlong2/latlong.dart';
import '../models/usuario.dart';
import '../models/servicio.dart';
import '../models/vehiculo.dart';
import '../models/evento_vial.dart';
import '../models/punto_geo.dart';

/// In-memory data layer used while the real backend is wired up.
/// Singleton, deterministic-ish, broadcasts changes through a stream so
/// the three apps behave like a single live system.
class MockDataService {
  MockDataService._internal() {
    _seed();
  }
  static final MockDataService instance = MockDataService._internal();

  final List<Usuario> _usuarios = [];
  final List<Servicio> _servicios = [];
  final List<Vehiculo> _vehiculos = [];
  final List<EventoVial> _eventos = [];

  List<Usuario> get usuarios => List.unmodifiable(_usuarios);
  List<Servicio> get servicios => List.unmodifiable(_servicios);
  List<Vehiculo> get vehiculos => List.unmodifiable(_vehiculos);
  List<EventoVial> get eventos => List.unmodifiable(_eventos);

  Usuario get clienteDemo => _usuarios.firstWhere((u) => u.perfil == TipoPerfil.cliente);
  Usuario get conductorDemo => _usuarios.firstWhere((u) => u.perfil == TipoPerfil.conductor);
  Usuario get adminDemo => _usuarios.firstWhere((u) => u.perfil == TipoPerfil.admin);

  List<Usuario> get conductores =>
      _usuarios.where((u) => u.perfil == TipoPerfil.conductor).toList();

  void _seed() {
    final now = DateTime.now();

    _usuarios.addAll([
      Usuario(
        id: 'u1',
        perfil: TipoPerfil.cliente,
        nombre: 'Mateo Rivas',
        correo: 'mateo@oltvi.app',
        telefono: '+54 9 11 5555 0001',
        nivel: NivelUsuario.experto,
        puntos: 1240,
        rating: 4.9,
        verificado: true,
        fechaRegistro: now.subtract(const Duration(days: 320)),
      ),
      Usuario(
        id: 'u2',
        perfil: TipoPerfil.conductor,
        nombre: 'Lucía Ferreyra',
        correo: 'lucia@oltvi.app',
        telefono: '+54 9 11 5555 0002',
        nivel: NivelUsuario.lider,
        puntos: 3870,
        rating: 4.95,
        verificado: true,
        fechaRegistro: now.subtract(const Duration(days: 480)),
      ),
      Usuario(
        id: 'u3',
        perfil: TipoPerfil.conductor,
        nombre: 'Diego Salinas',
        correo: 'diego@oltvi.app',
        telefono: '+54 9 11 5555 0003',
        nivel: NivelUsuario.experto,
        puntos: 1980,
        rating: 4.87,
        verificado: true,
        fechaRegistro: now.subtract(const Duration(days: 200)),
      ),
      Usuario(
        id: 'u4',
        perfil: TipoPerfil.conductor,
        nombre: 'Camila Ortega',
        correo: 'camila@oltvi.app',
        telefono: '+54 9 11 5555 0004',
        nivel: NivelUsuario.colaborador,
        puntos: 720,
        rating: 4.78,
        verificado: true,
        fechaRegistro: now.subtract(const Duration(days: 90)),
      ),
      Usuario(
        id: 'u5',
        perfil: TipoPerfil.admin,
        nombre: 'Centro de Mando OLTVI',
        correo: 'mando@oltvi.app',
        telefono: '+54 9 11 5555 0000',
        nivel: NivelUsuario.leyenda,
        puntos: 99999,
        rating: 5.0,
        verificado: true,
        fechaRegistro: now.subtract(const Duration(days: 720)),
      ),
    ]);

    _vehiculos.addAll([
      Vehiculo(
        id: 'v1',
        marca: 'Toyota',
        modelo: 'Corolla',
        anio: 2024,
        patente: 'AE 312 KP',
        color: 'Gris',
        tipo: TipoVehiculo.auto,
        capacidadPasajeros: 4,
        idConductor: 'u2',
        vencimientoSeguro: now.add(const Duration(days: 180)),
        vencimientoVtv: now.add(const Duration(days: 90)),
        consumoPromedioKmL: 14.5,
      ),
      Vehiculo(
        id: 'v2',
        marca: 'Renault',
        modelo: 'Kangoo',
        anio: 2023,
        patente: 'AC 887 LM',
        color: 'Blanco',
        tipo: TipoVehiculo.utilitario,
        capacidadKg: 650,
        idConductor: 'u3',
        vencimientoSeguro: now.add(const Duration(days: 45)),
        vencimientoVtv: now.add(const Duration(days: 10)),
        consumoPromedioKmL: 11.2,
      ),
      Vehiculo(
        id: 'v3',
        marca: 'Honda',
        modelo: 'Wave',
        anio: 2025,
        patente: 'A 1234 BC',
        color: 'Negro',
        tipo: TipoVehiculo.moto,
        capacidadKg: 15,
        idConductor: 'u4',
        vencimientoSeguro: now.add(const Duration(days: 300)),
        vencimientoVtv: now.add(const Duration(days: 200)),
        consumoPromedioKmL: 38.0,
      ),
    ]);

    // Seed servicios around Buenos Aires
    const baLat = -34.6037;
    const baLng = -58.3816;
    final rnd = Random(42);

    for (var i = 0; i < 8; i++) {
      final dx = (rnd.nextDouble() - 0.5) * 0.08;
      final dy = (rnd.nextDouble() - 0.5) * 0.08;
      final dx2 = (rnd.nextDouble() - 0.5) * 0.12;
      final dy2 = (rnd.nextDouble() - 0.5) * 0.12;
      final origen = PuntoGeo(
        lat: baLat + dx,
        lng: baLng + dy,
        nombre: _origenes[i % _origenes.length],
      );
      final destino = PuntoGeo(
        lat: baLat + dx2,
        lng: baLng + dy2,
        nombre: _destinos[i % _destinos.length],
      );
      final distKm = _distanceKm(origen.toLatLng(), destino.toLatLng());
      final tipo = TipoServicio.values[i % 3];
      final estado = i < 2
          ? EstadoServicio.enRuta
          : i < 4
              ? EstadoServicio.asignado
              : i < 6
                  ? EstadoServicio.entregado
                  : EstadoServicio.solicitado;
      _servicios.add(Servicio(
        id: 's${1000 + i}',
        tipo: tipo,
        nivel: NivelServicio.values[i % 4],
        origen: origen,
        destino: destino,
        idCliente: 'u1',
        idConductor: estado == EstadoServicio.solicitado ? null : 'u${2 + (i % 3)}',
        nombreConductor: estado == EstadoServicio.solicitado
            ? null
            : _usuarios[1 + (i % 3)].nombre,
        matriculaVehiculo: estado == EstadoServicio.solicitado
            ? null
            : _vehiculos[i % 3].patente,
        estado: estado,
        precio: 850 + distKm * 320 + rnd.nextInt(500),
        distanciaKm: distKm,
        tiempoEstimadoMin: (distKm * 3.5).round() + 4,
        fechaCreacion: now.subtract(Duration(minutes: 10 + i * 25)),
      ));
    }

    _eventos.addAll([
      EventoVial(
        id: 'e1',
        tipo: TipoEventoVial.bache,
        ubicacion: const PuntoGeo(lat: -34.6020, lng: -58.3795, nombre: 'Av. 9 de Julio'),
        descripcion: 'Bache profundo en carril central, esquivar derecha',
        idReportador: 'u3',
        fechaReporte: now.subtract(const Duration(hours: 2)),
        estado: EstadoEvento.confirmado,
        verificaciones: 8,
        rechazos: 1,
        iaScore: 0.94,
      ),
      EventoVial(
        id: 'e2',
        tipo: TipoEventoVial.obra,
        ubicacion: const PuntoGeo(lat: -34.6105, lng: -58.3920, nombre: 'San Telmo'),
        descripcion: 'Obra municipal, calle reducida a un carril',
        idReportador: 'u2',
        fechaReporte: now.subtract(const Duration(hours: 6)),
        estado: EstadoEvento.confirmado,
        verificaciones: 12,
        rechazos: 0,
        iaScore: 0.98,
      ),
      EventoVial(
        id: 'e3',
        tipo: TipoEventoVial.accidente,
        ubicacion: const PuntoGeo(lat: -34.5980, lng: -58.3700, nombre: 'Puerto Madero'),
        descripcion: 'Accidente leve, asistencia en camino',
        idReportador: 'u4',
        fechaReporte: now.subtract(const Duration(minutes: 18)),
        estado: EstadoEvento.verificando,
        verificaciones: 3,
        rechazos: 0,
        iaScore: 0.81,
      ),
      EventoVial(
        id: 'e4',
        tipo: TipoEventoVial.trafico,
        ubicacion: const PuntoGeo(lat: -34.5870, lng: -58.4000, nombre: 'Palermo'),
        descripcion: 'Congestión densa, evitar entre 18-20hs',
        idReportador: 'u3',
        fechaReporte: now.subtract(const Duration(minutes: 45)),
        estado: EstadoEvento.confirmado,
        verificaciones: 22,
        rechazos: 2,
        iaScore: 0.91,
      ),
    ]);
  }

  static const _origenes = [
    'Av. Corrientes 1500',
    'Plaza de Mayo',
    'Caballito',
    'Belgrano R',
    'Recoleta',
    'Palermo Soho',
    'Villa Crespo',
    'San Telmo',
  ];
  static const _destinos = [
    'Aeroparque',
    'Retiro',
    'Puerto Madero',
    'Núñez',
    'Almagro',
    'Microcentro',
    'Once',
    'Boedo',
  ];

  double _distanceKm(LatLng a, LatLng b) {
    const distance = Distance();
    return distance(a, b) / 1000.0;
  }

  Servicio crear(Servicio s) {
    _servicios.insert(0, s);
    return s;
  }

  void actualizar(Servicio s) {
    final i = _servicios.indexWhere((x) => x.id == s.id);
    if (i >= 0) _servicios[i] = s;
  }

  EventoVial crearEvento(EventoVial e) {
    _eventos.insert(0, e);
    return e;
  }

  void actualizarEvento(EventoVial e) {
    final i = _eventos.indexWhere((x) => x.id == e.id);
    if (i >= 0) _eventos[i] = e;
  }
}

File: android/core/src/main/kotlin/com/oltvi/core/data/services/MockDataService.kt
package com.oltvi.core.data.services

import com.oltvi.core.data.models.ConductorDisponible
import com.oltvi.core.data.models.EstadoServicio
import com.oltvi.core.data.models.EstadoVehiculo
import com.oltvi.core.data.models.NivelServicio
import com.oltvi.core.data.models.NivelUsuario
import com.oltvi.core.data.models.PuntoGeo
import com.oltvi.core.data.models.ResultadoMatchmaking
import com.oltvi.core.data.models.ResultadoPrecio
import com.oltvi.core.data.models.RutaOptimizada
import com.oltvi.core.data.models.Servicio
import com.oltvi.core.data.models.SolicitudViajeResultado
import com.oltvi.core.data.models.TipoPerfil
import com.oltvi.core.data.models.TipoServicio
import com.oltvi.core.data.models.TipoVehiculo
import com.oltvi.core.data.models.Usuario
import com.oltvi.core.data.models.Vehiculo
import java.time.Instant
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.abs
import kotlin.math.pow
import kotlin.math.sqrt

// Buenos Aires default coordinates
private const val BA_LAT = -34.6037
private const val BA_LNG = -58.3816

@Singleton
class MockDataService @Inject constructor() {

    // ── Predefined conductors ──────────────────────────────────────────────

    private val conductoresMock = listOf(
        DriverDef("driver-001", "Carlos Rodríguez", 4.8, "Toyota", "Corolla", 2022, "AB123CD", "Gris"),
        DriverDef("driver-002", "María González", 4.7, "Honda", "City", 2021, "EF456GH", "Blanco"),
        DriverDef("driver-003", "Diego Fernández", 4.9, "Volkswagen", "Polo", 2023, "IJ789KL", "Negro"),
        DriverDef("driver-004", "Lucía Martínez", 4.5, "Chevrolet", "Cruze", 2020, "MN012OP", "Plateado"),
        DriverDef("driver-005", "Andrés López", 4.6, "Ford", "Focus", 2021, "QR345ST", "Azul"),
        DriverDef("driver-006", "Valentina García", 4.8, "Nissan", "Sentra", 2022, "UV678WX", "Rojo"),
        DriverDef("driver-007", "Pablo Sánchez", 4.2, "Renault", "Logan", 2019, "YZ901AB", "Beige"),
        DriverDef("driver-008", "Camila Díaz", 4.7, "Toyota", "Yaris", 2023, "CD234EF", "Blanco Perla")
    )

    // ── Mock users ─────────────────────────────────────────────────────────

    private val usuariosMock = listOf(
        Usuario(
            id = "user-001",
            perfil = TipoPerfil.CLIENTE,
            nombre = "Alejandro Torres",
            correo = "alejandro.torres@gmail.com",
            telefono = "+54 11 5555-1234",
            nivel = NivelUsuario.EXPERTO,
            puntos = 3750,
            rating = 4.8,
            verificado = true
        ),
        Usuario(
            id = "user-002",
            perfil = TipoPerfil.CLIENTE,
            nombre = "Sofía Romero",
            correo = "sofia.romero@outlook.com",
            telefono = "+54 11 4444-5678",
            nivel = NivelUsuario.COLABORADOR,
            puntos = 820,
            rating = 4.6,
            verificado = true
        ),
        Usuario(
            id = "user-003",
            perfil = TipoPerfil.CLIENTE,
            nombre = "Martín Pérez",
            correo = "martin.perez@gmail.com",
            telefono = "+54 11 3333-9012",
            nivel = NivelUsuario.INICIADO,
            puntos = 150,
            rating = 4.9,
            verificado = false
        ),
        Usuario(
            id = "user-004",
            perfil = TipoPerfil.ADMIN,
            nombre = "Laura Vega",
            correo = "laura.vega@oltvi.com",
            telefono = "+54 11 2222-3456",
            nivel = NivelUsuario.LEYENDA,
            puntos = 25000,
            rating = 5.0,
            verificado = true
        ),
        Usuario(
            id = "user-005",
            perfil = TipoPerfil.CLIENTE,
            nombre = "Facundo Méndez",
            correo = "facundo.mendez@yahoo.com.ar",
            telefono = "+54 11 1111-7890",
            nivel = NivelUsuario.LIDER,
            puntos = 7200,
            rating = 4.7,
            verificado = true
        )
    )

    // ── Public API ─────────────────────────────────────────────────────────

    /**
     * Returns all 8 mock conductors with realistic distances/ETAs computed
     * from the given lat/lng reference point.
     */
    fun getDriversNear(lat: Double, lng: Double): List<ConductorDisponible> {
        // Fixed offsets to make distances deterministic per driver
        val offsets = listOf(
            Pair(0.0032, -0.0041),
            Pair(-0.0055, 0.0028),
            Pair(0.0071, 0.0019),
            Pair(-0.0024, -0.0062),
            Pair(0.0088, -0.0033),
            Pair(-0.0011, 0.0075),
            Pair(0.0049, 0.0052),
            Pair(-0.0067, -0.0018)
        )

        return conductoresMock.mapIndexed { i, def ->
            val (dLat, dLng) = offsets[i]
            val driverLat = lat + dLat
            val driverLng = lng + dLng
            val distanciaMetros = calcularDistanciaMetros(
                lat, lng, driverLat, driverLng
            )
            val etaMin = (distanciaMetros / 400.0).toInt().coerceIn(2, 18)

            ConductorDisponible(
                idConductor = def.id,
                nombre = def.nombre,
                rating = def.rating,
                distanciaMetros = distanciaMetros,
                tiempoEstimadoMin = etaMin,
                vehiculo = Vehiculo(
                    id = "veh-${def.id}",
                    marca = def.marca,
                    modelo = def.modelo,
                    anio = def.anio,
                    patente = def.patente,
                    color = def.color,
                    tipo = TipoVehiculo.AUTO,
                    estado = EstadoVehiculo.OPERATIVO,
                    capacidadPasajeros = 4,
                    idConductor = def.id
                ),
                posicion = PuntoGeo(driverLat, driverLng, def.nombre),
                scoreFit = 0.0
            )
        }.sortedBy { it.distanciaMetros }
    }

    /**
     * Creates a service request with Haversine-based distance, price and ETA.
     */
    fun createServiceRequest(
        origen: PuntoGeo,
        destino: PuntoGeo,
        tipo: TipoServicio
    ): Servicio {
        val distanciaKm = calcularDistanciaKm(origen, destino).coerceAtLeast(0.5)
        val precio = (800.0 + 120.0 * distanciaKm) * tipo.priceMultiplierFor()
        val etaMin = (distanciaKm / 30.0 * 60.0).toInt().coerceAtLeast(2)
        val conductor = conductoresMock.first()

        return Servicio(
            id = UUID.randomUUID().toString(),
            tipo = tipo,
            nivel = NivelServicio.ESTANDAR,
            origen = origen,
            destino = destino,
            idCliente = "user-001",
            idConductor = conductor.id,
            nombreConductor = conductor.nombre,
            matriculaVehiculo = conductor.patente,
            estado = EstadoServicio.SOLICITADO,
            precio = precio,
            distanciaKm = distanciaKm,
            tiempoEstimadoMin = etaMin
        )
    }

    fun getUserById(id: String): Usuario? = usuariosMock.find { it.id == id }

    /**
     * Returns 5 active services in different states (realistic Buenos Aires data).
     */
    fun getActiveServices(): List<Servicio> {
        val now = Instant.now()
        val conductores = conductoresMock
        val estados = listOf(
            EstadoServicio.SOLICITADO,
            EstadoServicio.EN_CAMINO,
            EstadoServicio.EN_RUTA,
            EstadoServicio.LLEGANDO,
            EstadoServicio.ENTREGADO
        )
        val rutas = listOf(
            Pair(PuntoGeo(BA_LAT, BA_LNG, "Microcentro"), PuntoGeo(-34.5797, -58.4325, "Palermo")),
            Pair(PuntoGeo(-34.5875, -58.3938, "Recoleta"), PuntoGeo(-34.6214, -58.3731, "San Telmo")),
            Pair(PuntoGeo(-34.5607, -58.4531, "Belgrano"), PuntoGeo(BA_LAT, BA_LNG, "Microcentro")),
            Pair(PuntoGeo(-34.6183, -58.4417, "Caballito"), PuntoGeo(-34.5534, -58.4597, "Belgrano")),
            Pair(PuntoGeo(-34.6230, -58.3890, "San Telmo"), PuntoGeo(-34.5885, -58.4260, "Palermo Soho"))
        )

        return rutas.mapIndexed { i, (origen, destino) ->
            val c = conductores[i]
            val distanciaKm = calcularDistanciaKm(origen, destino)
            val precio = 800.0 + 120.0 * distanciaKm
            Servicio(
                id = "active-00${i + 1}",
                tipo = TipoServicio.PASAJERO,
                nivel = NivelServicio.ESTANDAR,
                origen = origen,
                destino = destino,
                idCliente = "user-00${(i % 5) + 1}",
                idConductor = c.id,
                nombreConductor = c.nombre,
                matriculaVehiculo = c.patente,
                estado = estados[i],
                precio = precio,
                distanciaKm = distanciaKm,
                tiempoEstimadoMin = (distanciaKm / 30.0 * 60.0).toInt().coerceAtLeast(3),
                fechaCreacion = now.minusSeconds((i * 900).toLong())
            )
        }
    }

    // ── Backward-compat helpers used by old code ───────────────────────────

    fun getMockUser(): Usuario = usuariosMock.first()

    fun buildMockSolicitudResultado(destino: PuntoGeo): SolicitudViajeResultado {
        val drivers = getDriversNear(BA_LAT, BA_LNG)
        val elegido = drivers.first()
        val distanciaKm = calcularDistanciaKm(
            PuntoGeo(BA_LAT, BA_LNG), destino
        ).coerceAtLeast(0.5)
        val precio = 800.0 + 120.0 * distanciaKm
        return SolicitudViajeResultado(
            matchmaking = ResultadoMatchmaking(
                conductorElegido = elegido,
                alternativas = drivers.drop(1).take(2),
                explicacionIA = "Seleccioné a ${elegido.nombre} por su rating de ${elegido.rating}⭐ " +
                        "y cercanía de ${"%.0f".format(elegido.distanciaMetros)}m — la mejor combinación disponible.",
                confianza = 0.94f
            ),
            precio = ResultadoPrecio(
                precioFinal = precio,
                desglose = mapOf(
                    "Precio base" to 800.0,
                    "Distancia (${"%.1f".format(distanciaKm)} km)" to (120.0 * distanciaKm),
                    "Factor demanda (×1.1)" to precio * 0.1,
                    "Descuento usuario" to -(precio * 0.05)
                ),
                explicacionIA = "Precio de \$${"%.0f".format(precio)} ARS para ${"%.1f".format(distanciaKm)} km con demanda normal.",
                factorDemanda = 1.1,
                factorClima = "Despejado"
            ),
            ruta = RutaOptimizada(
                puntos = listOf(
                    PuntoGeo(BA_LAT, BA_LNG, "Origen"),
                    PuntoGeo(BA_LAT + 0.005, BA_LNG + 0.003, "Punto intermedio"),
                    destino
                ),
                etaMin = elegido.tiempoEstimadoMin + (distanciaKm / 30.0 * 60.0).toInt().coerceAtLeast(3),
                distanciaKm = distanciaKm,
                trafico = "Normal"
            )
        )
    }

    fun getEventosViales(): List<EventoVial> {
        val now = java.time.Instant.now()
        return listOf(
            EventoVial("ev1", TipoEventoVial.BACHE, "Bache profundo en la calzada", PuntoGeo(-34.6037, -58.3816, "Av. Corrientes"), "usr1", now.minusSeconds(1800), EstadoEventoVial.CONFIRMADO, SeveridadAlerta.MEDIA),
            EventoVial("ev2", TipoEventoVial.OBRA, "Obra de gas, carril izquierdo cortado", PuntoGeo(-34.6100, -58.3900, "Santa Fe y Callao"), "usr2", now.minusSeconds(3600), EstadoEventoVial.REPORTADO, SeveridadAlerta.ALTA),
            EventoVial("ev3", TipoEventoVial.ACCIDENTE, "Choque entre dos vehículos. Tránsito lento.", PuntoGeo(-34.5980, -58.3750, "Autopista 25 de Mayo"), "usr3", now.minusSeconds(600), EstadoEventoVial.VERIFICANDO, SeveridadAlerta.ALTA),
            EventoVial("ev4", TipoEventoVial.CORTE_TOTAL, "Marcha política — calle cerrada", PuntoGeo(-34.6150, -58.3700, "9 de Julio y Diagonal Norte"), "usr1", now.minusSeconds(7200), EstadoEventoVial.CONFIRMADO, SeveridadAlerta.CRITICA),
            EventoVial("ev5", TipoEventoVial.INUNDACION, "Agua acumulada, tránsito dificultoso", PuntoGeo(-34.6200, -58.4000, "Av. San Juan"), "usr4", now.minusSeconds(2400), EstadoEventoVial.REPORTADO, SeveridadAlerta.MEDIA),
            EventoVial("ev6", TipoEventoVial.OTRO, "Semáforo fuera de servicio", PuntoGeo(-34.6050, -58.3850, "Av. Rivadavia y Pueyrredón"), "usr2", now.minusSeconds(900), EstadoEventoVial.CONFIRMADO, SeveridadAlerta.BAJA),
        )
    }

    // ── Private helpers ────────────────────────────────────────────────────

    private fun calcularDistanciaMetros(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val dLat = (lat1 - lat2) * 111000.0
        val dLng = (lng1 - lng2) * 111000.0 * 0.82
        return sqrt(dLat.pow(2) + dLng.pow(2))
    }

    private fun calcularDistanciaKm(a: PuntoGeo, b: PuntoGeo): Double {
        val dLat = (a.lat - b.lat) * 111.0
        val dLng = (a.lng - b.lng) * 111.0 * 0.82
        return sqrt(dLat.pow(2) + dLng.pow(2))
    }

    private fun TipoServicio.priceMultiplierFor(): Double = when (this) {
        TipoServicio.PASAJERO -> 1.0
        TipoServicio.MENSAJERIA -> 0.8
        TipoServicio.CARGA -> 1.5
        TipoServicio.VIAL -> 0.5
    }
}

private data class DriverDef(
    val id: String,
    val nombre: String,
    val rating: Double,
    val marca: String,
    val modelo: String,
    val anio: Int,
    val patente: String,
    val color: String
)
